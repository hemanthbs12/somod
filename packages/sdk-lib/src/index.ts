export * from "./tasks/packageJson";

export * from "./tasks/ignoreFiles";

export * from "./tasks/tsConfigBuildJson";

export * from "./tasks/init";

export * from "./tasks/build";

export * from "./tasks/link";

export * from "./tasks/dev";

export * from "./tasks/serverless";

export * from "./utils/constants";
