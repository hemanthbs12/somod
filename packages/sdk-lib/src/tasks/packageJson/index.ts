export {
  update as updatePackageJson,
  updateSodaruModuleKey as updateSodaruModuleKeyInPackageJson
} from "./update";
export { save as savePackageJson } from "./save";
export { validate as validatePackageJson } from "./validate";
