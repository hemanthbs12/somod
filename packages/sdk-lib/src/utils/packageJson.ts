import { join, normalize } from "path";
import { file_packageJson } from "./constants";
import {
  readJsonFileStore,
  updateJsonFileStore,
  unixStylePath
} from "@solib/cli-base";

export const read = async (dir: string): Promise<Record<string, unknown>> => {
  const packageJsonPath = join(dir, file_packageJson);
  const packageJsonContent = await readJsonFileStore(packageJsonPath);
  return packageJsonContent;
};

export const update = (
  dir: string,
  packageJson: Record<string, unknown>
): void => {
  const packageJsonPath = join(dir, file_packageJson);
  updateJsonFileStore(packageJsonPath, packageJson);
};

export const packageJsonPath = (dir: string): string => {
  return unixStylePath(normalize(join(dir, file_packageJson)));
};
