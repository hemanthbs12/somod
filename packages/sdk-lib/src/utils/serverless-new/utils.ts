import { createHash } from "crypto";

export const getNodeRuntimeVersion = () => {
  return process.env.SOMOD_SERVERLESS_NODEJS_VERSION || "16";
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  somodResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + somodResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  somodResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "somod${stackId}${moduleHash}${somodResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        somodResourceName: somodResourceName
      }
    ]
  };
};

export const getSAMOutputName = (exportParameterName: string) => {
  return "o" + Buffer.from(exportParameterName).toString("hex");
};

export const getParameterNameFromSAMOutputName = (samOutputName: string) => {
  return Buffer.from(samOutputName.substring(1), "hex").toString();
};
