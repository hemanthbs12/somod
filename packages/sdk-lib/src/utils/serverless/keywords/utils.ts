import { cloneDeep, isArray, isPlainObject, isUndefined } from "lodash";
import {
  KeywordAll,
  KeywordSLPDependsOn,
  KeywordSLPFunction,
  KeywordSLPRef,
  KeywordSLPRefParameter,
  KeywordSLPRefResourceName,
  SLPDependsOn,
  SLPFunction,
  SLPKeyword,
  SLPRef,
  SLPRefParameter,
  SLPRefResourceName,
  SLPTemplate
} from "../types";

const getSLPKeywordPaths = (chunk: unknown): SLPTemplate["keywordPaths"] => {
  const keywordPaths = Object.fromEntries(
    KeywordAll.map(keyword => [keyword, []])
  ) as SLPTemplate["keywordPaths"];

  if (isPlainObject(chunk)) {
    KeywordAll.forEach(keyword => {
      if (!isUndefined(chunk[keyword])) {
        keywordPaths[keyword].push([]);
      }
    });

    Object.keys(chunk).forEach(key => {
      const childKeywordPaths = getSLPKeywordPaths(chunk[key]);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(key);
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  } else if (isArray(chunk)) {
    chunk.forEach((arrayItem, index) => {
      const childKeywordPaths = getSLPKeywordPaths(arrayItem);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(index + "");
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  }
  return keywordPaths;
};

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getSLPKeywordPaths(slpTemplate.Resources);
  slpTemplate.keywordPaths = slpKeywords;
};

export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  paths: string[]
): T => {
  let keyword: unknown = slpTemplate.Resources;
  paths.forEach(path => {
    keyword = keyword[path];
  });
  return cloneDeep(keyword) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  paths: string[],
  newValue: unknown
): void => {
  let keyword: unknown = slpTemplate.Resources;
  const _paths = [...paths];
  const lastPath = _paths.pop();
  _paths.forEach(path => {
    keyword = keyword[path];
  });
  keyword[lastPath] = newValue;
};

/**
 * some of the slp keywords references the modulename , by default the module name is the current module of the SLP Template
 * fill the default value of modulename
 */
export const updateCurrentModuleInSLPTemplate = (
  module: string,
  slpTemplate: SLPTemplate
): void => {
  type TypeKeywordsToUpdate =
    | typeof KeywordSLPDependsOn
    | typeof KeywordSLPRef
    | typeof KeywordSLPRefParameter
    | typeof KeywordSLPRefResourceName
    | typeof KeywordSLPFunction;

  const keywordsToUpdate: TypeKeywordsToUpdate[] = [
    KeywordSLPDependsOn,
    KeywordSLPRef,
    KeywordSLPRefParameter,
    KeywordSLPRefResourceName,
    KeywordSLPFunction
  ];

  keywordsToUpdate.forEach(keyword => {
    slpTemplate.keywordPaths[keyword].forEach(keywordPaths => {
      const keywordValue = getSLPKeyword<
        | SLPDependsOn
        | SLPRef
        | SLPRefParameter
        | SLPRefResourceName
        | SLPFunction
      >(slpTemplate, keywordPaths);
      if (keyword == KeywordSLPFunction) {
        keywordValue[keyword] = {
          module,
          function: keywordValue[keyword]
        };
      } else if (keywordValue[keyword].module === undefined) {
        keywordValue[keyword].module = module;
      }
      replaceSLPKeyword(slpTemplate, keywordPaths, keywordValue);
    });
  });
};
