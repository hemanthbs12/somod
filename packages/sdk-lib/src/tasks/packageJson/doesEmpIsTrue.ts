import { key_emp } from "../../utils/constants";
import { read, packageJsonPath } from "../../utils/packageJson";

export const doesEmpIsTrue = async (dir: string): Promise<void> => {
  const packageJson = await read(dir);

  if (packageJson[key_emp] !== true) {
    throw new Error(`${key_emp} must be true in ${packageJsonPath(dir)}`);
  }
};
