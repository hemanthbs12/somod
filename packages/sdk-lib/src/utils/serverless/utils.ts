import { createHash } from "crypto";
import { getKeyword, getKeywordPaths, replaceKeyword } from "../keywords";
import { KeywordAll, SLPKeyword, SLPTemplate } from "./types";

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getKeywordPaths(
    slpTemplate.original.Resources,
    KeywordAll
  );
  slpTemplate.keywordPaths = slpKeywords;
};

export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  path: string[]
): T => {
  // always return the keyword value from original
  return getKeyword(slpTemplate.original.Resources, path) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  path: string[],
  newValue: unknown
): void => {
  replaceKeyword(slpTemplate.Resources, path, newValue);
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getParameterSpaceResourceLogicalId = (
  parameterSpace: string
): string => {
  return "p" + parameterSpace;
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  slpResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + slpResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  slpResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "slp${stackId}${moduleHash}${slpResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        slpResourceName
      }
    ]
  };
};
