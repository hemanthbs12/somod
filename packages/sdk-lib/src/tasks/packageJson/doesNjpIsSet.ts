import { key_njp, cli_version_regex } from "../../utils/constants";
import { read, packageJsonPath } from "../../utils/packageJson";

export const doesNjpIsSet = async (dir: string): Promise<void> => {
  const packageJson = await read(dir);

  if (!cli_version_regex.test(packageJson[key_njp] as string)) {
    throw new Error(
      `${key_njp} must match '${cli_version_regex}' pattern in ${packageJsonPath(
        dir
      )}`
    );
  }
};
