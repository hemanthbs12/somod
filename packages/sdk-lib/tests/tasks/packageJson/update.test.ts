import { createFiles, createTempDir, deleteDir } from "../../utils";
import { updatePackageJson } from "../../../src";
import { read as readPackageJson } from "../../../src/utils/packageJson";

describe("Test Task updatePackageJson", () => {
  let dir: string = null;

  beforeEach(() => {
    dir = createTempDir();
  });

  afterEach(() => {
    deleteDir(dir);
  });

  const allRightPackageJson = {
    name: "@my-scope/my-module",
    version: "1.0.1",
    description: "",
    module: "build/lib/index.js",
    typings: "build/lib/index.d.ts",
    files: ["build"],
    sideEffects: false
  };

  const njp = {
    njp: "1.2.3"
  };

  const testData: [string, Record<string, unknown>][] = [
    ["empty", {}],
    [
      "normal",
      {
        name: "@my-scope/my-module",
        version: "1.0.1",
        description: ""
      }
    ],
    [
      "no sodaru module key",
      {
        ...allRightPackageJson
      }
    ],
    [
      "njp and somod",
      {
        ...allRightPackageJson,
        ...njp,
        somod: "1.2.3"
      }
    ],
    [
      "only njp",
      {
        ...allRightPackageJson,
        ...njp
      }
    ],
    [
      "wrong module",
      {
        ...allRightPackageJson,
        ...njp,
        module: "build/lib/index.ts"
      }
    ],
    [
      "wrong typings",
      {
        ...allRightPackageJson,
        ...njp,
        typings: "dist/lib/index.ts"
      }
    ],
    [
      "wrong files",
      {
        ...allRightPackageJson,
        ...njp,
        files: []
      }
    ],
    [
      "wrong sideEffects",
      {
        ...allRightPackageJson,
        ...njp,
        sideEffects: true
      }
    ],
    [
      "not-allowed keys",
      {
        ...allRightPackageJson,
        ...njp,
        main: "build/lib/index.js",
        "jsnext:main": "build/lib/index.js",
        type: "esm"
      }
    ],
    [
      "only somod set for njp module",
      { ...allRightPackageJson, somod: "1.2.3" }
    ]
  ];

  test.each(testData)("with %s", async (title, content) => {
    createFiles(dir, { "package.json": JSON.stringify(content) });

    await updatePackageJson(dir, "njp");
    const result = await readPackageJson(dir);
    expect(result).toMatchSnapshot({
      njp: expect.any(String)
    });
  });

  test("with type = slp", async () => {
    createFiles(dir, { "package.json": "{}" });

    await updatePackageJson(dir, "slp");
    const result = await readPackageJson(dir);
    expect(result).toMatchSnapshot({
      slp: expect.any(String)
    });
  });

  test("with type = somod", async () => {
    createFiles(dir, { "package.json": "{}" });

    await updatePackageJson(dir, "somod");
    const result = await readPackageJson(dir);
    expect(result).toMatchSnapshot({
      somod: expect.any(String)
    });
  });
});
