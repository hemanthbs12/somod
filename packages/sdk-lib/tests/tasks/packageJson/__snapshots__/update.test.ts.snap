// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Task updatePackageJson with empty 1`] = `
Object {
  "description": undefined,
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": undefined,
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": undefined,
}
`;

exports[`Test Task updatePackageJson with njp and somod 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with no sodaru module key 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with normal 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with not-allowed keys 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with only njp 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with only somod set for njp module 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with wrong files 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with wrong module 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with wrong sideEffects 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;

exports[`Test Task updatePackageJson with wrong typings 1`] = `
Object {
  "description": "",
  "files": Array [
    "build",
  ],
  "module": "build/index.js",
  "name": "@my-scope/my-module",
  "njp": Any<String>,
  "sideEffects": false,
  "typings": "build/index.d.ts",
  "version": "1.0.1",
}
`;
